hs/src/Reach/AST/SL.hs

```diff
@@ -160,6 +160,38 @@ data SLVal
   | SLV_Deprecated Deprecation SLVal
   deriving (Eq, Generic)
 
+-- | Equivalence operation on flattened or simplified structures.
+class Equiv a where
+  equiv :: a -> a -> Bool
+
+instance Equiv SLVal where
+  equiv (SLV_Null _ s1) (SLV_Null _ s2) = s1 == s2
+  equiv (SLV_Bool _ b1) (SLV_Bool _ b2) = b1 == b2
+  equiv (SLV_Int _ i1) (SLV_Int _ i2) = i1 == i2
+  equiv (SLV_Bytes _ v1) (SLV_Bytes _ v2) = v1 == v2
+  equiv (SLV_Array _ t xs) (SLV_Array _ t2 ys) = t == t2 && xs == ys
+  equiv (SLV_Tuple _ v1) (SLV_Tuple _ v2) = v1 == v2
+  equiv (SLV_Struct _ xs) (SLV_Struct _ ys) = xs == ys
+  equiv (SLV_DLC d1) (SLV_DLC d2) = d1 == d2
+  equiv (SLV_DLVar d1) (SLV_DLVar d2) = d1 == d2
```

Comment:
They may have different source location information, so you should implement equiv